### PR TITLE
Link to ev3python.com

### PR DESCRIPTION
**Pull request by ddemidov**

### CODE REVIEW

**Comment by ddemidov** [comment]
Pinging @ndward

**Comment by WasabiFan** [comment]
I like where this is going, but I think it should be trumpeted a bit more: his site isn't just a reference, but essentially an end-to-end guide for a beginner. What do you think about having it say something like "If you are a beginner with Python and would like extra support to guide you through using ev3dev with Python, please check out ..."? Maybe that sentence could be bolded too if it's at the end of the paragraph.

**Comment by WasabiFan** [comment]
By the way, thanks @ddemidov for starting the work on this! And of course thanks to @ndward for the great content!

I've gotten through reading a few pages in detail from the EV3 Python site, but I'm still struggling to find time to write down some meaningful feedback. I have no problem with these links being added before I can get to that though.

**Comment by ddemidov** [comment]
How about:

We feel that Python is the best option if you are just starting programming. The [ev3python.com](http://ev3python.com) by @ndward is an excellent guide through using ev3dev with Python.

**Comment by cavenel** [comment]
And thank you @ndward for the very nice video of what is doable with ev3dev and Python ;-)

**Comment by bmegli** [comment]
> We feel that Python is the best option if you are just starting programming

The question is - do we really feel it  😄 ? This is my personal preference, I don't like dynamically typed languages unless for really small tasks.

But I have nothing else to recommend. Javascript suffers from the same problem (dynamic typing). C/C++ may seem like a joke for a beginner.

I don't know google's Go language but I think I might want to give it a go if I were beginner, just like python. 

Who knows Go here? How about Go vs Python for beginner?

@kortschak what can you say us about Go. Is it suitable for beginners? Do you know python also? How about ease of ease for newcommers?

**Comment by bmegli** [comment]
Also I like having and giving options instead of pointing everyone to one way, like:

If you are just starting programming your best options may be Python or [the second super cool language]. With some overview what pros and cons each languages has.

**Comment by ddemidov** [comment]
I am a C++ addict, but I find myself using python more and more as a scripting language. C++ does the heavy lifting, while python is used as an interactive playground.

Although here, I think, its not about which language is better in the long run, more powerful, or just more fashionable; its about which language is easier to pick up by a 10-year old, and it seems Python really shines here.

But I can remove or reword the first sentence if people don't like it.

**Comment by bmegli** [comment]
> I am a C++ addict, but I find myself using python more and more as a scripting language. C++ does the heavy lifting, while python is used as an interactive playground.

Similiar, I am just using R as interactive playground also for it's features of data visualization.

If you feel that it's really easier to pick up by newcommers go for it.

Like I said I have nothing else to recommend. I don't know the second option that seems reasonable (Go) for newcommers.

**Comment by kortschak** [comment]
>@kortschak what can you say us about Go. Is it suitable for beginners? Do you know python also? How about ease of use for newcommers?

With some explanation of what a pointer is (see the link below), I think that Go should be easily understandable by a not-too-young child (ten seems reasonable in most cases). There is a good discussion [here](https://groups.google.com/d/topic/golang-nuts/FIRSDBehb3g/discussion) where someone was using Go as a teaching language for 11 year-olds.

Yes, I write in python occasionally, although I avoid it and usually interact with it when I translating something from it. The two major problems I have with it - that I don't with Go (except in some cases where people are avoiding sensible use of types) - are the use of meaningful whitespace to denote scope and body, and most significantly the absence of type annotation.

**Comment by bmegli** [comment]
@kortschak Yes, but would you recommend Go for newcommers along with Python? I don't only mean the language itself but the IDE, ease of use with ev3dev.

The newcomers will be struggling with:
- language syntax
- ev3dev API
- robotics concepts (motor modes etc)
- toolchain (how to copy the program to EV3? do I have to compile it on PC, can I do it on EV3, maybe I don't have to compile it at all)
- IDE if there is
- building hardware
- linux
- networking
- ...

This is a lot for a person to learn.

So speciffically language syntax, API, toolchain and IDE - do we have something to propose for every element here for Go? Is it possible to understand for newcommer struggling with everything that it's on this list?

**Comment by ddemidov** [comment]
> The newcomers will be struggling with:

Ok, then one advantage of python is that @ndward actually made the work of introducing all this to beginners on his site :).

**Comment by kortschak** [comment]
For Go there is a fair amount of learning material. The most common starting point would be [the tour](https://tour.golang.org/welcome/1), but also [Go by example](https://gobyexample.com/) - and I would like eventually to have a complete set of translations of the LME programs that come with the educational set as examples in `github.com/ev3go/ev3dev/examples/.

The ev3dev API is a constant, so it's not necessary to consider here, though I think the `github.com/ev3go/ev3dev` reflection of it should make it reasonably straightforward.

There is an IDE for Go ([LiteIDE](https://github.com/visualfc/liteide)) which I don't use because the build environment is trivial to set up and the build chain is very simple to use (I find installing python packages much more painful): for example, installing the ev3-specific ev3go package and all dependencies is just `go get github.com/ev3go/ev3` - this brings `github.com/ev3go/ev3dev` and all the non-standard imports (actually minimal in this case) along with it.

**Comment by ddemidov** [comment]
How about:

> If you are just starting programming ev3dev, take a look at [ev3python.com](http://ev3python.com) by @ndward. The site is intended to help users of the Lego EV3 Mindstorms to begin using the ev3dev Python programming language to control their robot. 

Note that this goes into the 'Python' section of the page.

**Comment by bmegli** [comment]
@kortschak 

> There is an IDE for Go (LiteIDE) which I don't use because the build environment is trivial to set up and the build chain is very simple to use (I find installing python packages much more painful)

Yes, I mean that (hypothetical) 10-year old is not going to compile from the command line, copy the file with scp, log with ssh and run the program. He need's IDE that will take care of that for him so that he only has to write the code (on his PC not EV3!) and click the "Run" button.

**Comment by kortschak** [comment]
I'm not convinced the situation is different to python. In both cases you need to edit the file locally and transfer an object to the ev3. In the case of python the compilation is presumably then done on the device while in the Go case it's done on the host (though can be done on the device as well if wanted). The file/object transfer is the same process, as is starting the program.

I could imagine though writing a plugin for LiteIDE or SublimeText to automate the build and object transfer.

How does the python case work - a quick look at the pages on ev3python.com dealing with interacting with code and the device don't look anymore friendly than what I describe above.

**Comment by ddemidov** [comment]
In case of python you could in principle both write and run the (almost unchanged) code on desktop PC: https://gist.github.com/ddemidov/63dab1dd1b91b493a7752c76e055b5c0.

But I think the biggest advantage is that python is an interpreted language and does not require the cross-compilation step which may be difficult to setup for uninitiated.

**Comment by kortschak** [comment]
Cross compilation is this: `GOARCH=arm GOARM=5l go build`. The build on the device is the same without the var prefix. I don't see the difference between an interpretted language and a compiled language to be really all that significant when compile times are short.

The RPC approach is interesting. I have to look at what that would entail.

**Comment by bmegli** [comment]
Guys, I think that my workflow for C/C++ is the simplest for newcomers 😄 !

You just create ev3dev project in CodeLite, compile it as usual, and run it as usual only it runs on EV3.

Let's make C/C++ language for newcommers... 

@kortschak
> I could imagine though writing a plugin for LiteIDE or SublimeText to automate the build and object transfer.

I don't think you would need a plugin for that if IDE supports post-build commands and you can specify run commands

Also transferring can be made simple by filesharing instead (either working on EV3 filesystem or EV3 having access to PC filesystem)

@ddemidov
>But I think the biggest advantage is that python is an interpreted language and does not require the cross-compilation step which may be difficult to setup for uninitiated.

From what I read Go has built in compiler so that you can use it similarly to interpreted languages.

@ddemidov +1 for uninitiated 😄

**Comment by WasabiFan** [comment]
Okay... In the time it took me to sleep, this thread has significantly grown 😆  Let's save the discussion on various languages for now and focus on getting this PR ready and merged.

File: docs/programming-languages.md [hunk 1]
```diff
@@ -17,6 +17,10 @@ Python is a high-level, general-purpose, interpreted scripting language. It is
 great for beginners, while also powerful for experienced coders.If you are new
 to programming and unsure which language to choose, Python is a great choice!
 
+We feel that Python is the best option if you are just starting programming.
+The [ev3python.com](http://ev3python.com) by @ndward is an excellent guide
```
Comment (WasabiFan) [inline docs/programming-languages.md]: I think this would flow more nicely if you were to remove "the" here.

Comment (ddemidov) [inline docs/programming-languages.md]: There is another proposal in the thread above (really, you should sleep less!):

> If you are just starting programming ev3dev, take a look at ev3python.com by @ndward. The site is intended to help users of the Lego EV3 Mindstorms to begin using the ev3dev Python programming language to control their robot.

Comment (WasabiFan) [inline docs/programming-languages.md]: Oh, I read that... I just didn't realize that this sentence was different 😪 I like it how I said I shouldn't code review at 5:00 in the morning yesterday so instead I reviewed this at 6:00.

Anyway, yes I like that one better :laughing: Although, the capitalization that we use (based on what is used upstream) is all-caps LEGO MINDSTORMS EV3. Sorry for making this so difficult.

Comment (ddemidov) [inline docs/programming-languages.md]: Thats ok :). See d3b612

Comment (ddemidov) [inline docs/programming-languages.md]: 694fa38

**Comment by WasabiFan** [comment]
Thanks!

**Comment by bmegli** [comment]
@kortschak 

> I'm not convinced the situation is different to python.

That's one of the reasons I didn't like the sentence:

> We feel that Python is the best option if you are just starting programming

But what @ddemidov chosen instead is great.

I meant that other options may be just as good (e.g. Go) so we shouldn't specifically sanctify python.

**Comment by ndward** [comment]
First, I thank @ddemidov for opening this discussion about a link to my site, [ev3python.com](url). Seeing the title, I at first assumed that the discussion would be about whether there should be a link to the site at all, and if so, where it should be placed. Instead, it has largely been about which (textual) language young beginner programmers should be encouraged to learn first. The discussion has focused on technical aspects of the language. I would suggest that Python is probably the best because even if is not technically superior it has the edge in conciseness, readability, popularity and, crucially, support. By 'popularity', I mean that schools and universities in the USA and UK have massively opted for Python. I'm guessing it's also more popular in the workplace. By 'support', I mean that there are far more materials out there to support young Python learners than young learners of other languages such as Go. Not only is Python probably the best textual language for beginners to learn, EV3 Python is also the best supported EV3 programming option, given that 50 pages of EV3 Python-specific help are available. If it is felt that Go is an equally good option of beginners, taking into account the beginner-level support that is available for standard Go and specifically for EV3 Go, then ev3dev.org could perfectly well state that either Go or Python would be great choices for beginners.

Getting back to what I thought was going to be the theme of this discussion, and assuming that the ev3dev team decides to link to my site, then the question remains of how the link should be worded and where it should be placed. I have been encouraged to put in a 'pull request' for a link to be included on the projects page but I think that's definitely a bad idea. The projects page is currently a showcase for very sophisticated individual programs in multiple languages but my site is almost the exact opposite - it's full of extremely short, very simple programs (dozens of them, not just one). There is almost no chance that beginner Python programmers would be looking through the projects page for help getting started, and even if they did they would probably not find the link since it would be buried amongst at least 17 other projects. 

My site is essentially a tutorial site so it could be suggested that a link should be placed on the tutorials page, but again there is very little chance of beginner Python programmers finding it there, buried amongst about 30 other tutorials.

For me it's obvious that the link should go in the Python paragraph on the Programming Languages page where every young Python programmer is likely to see it as part of the getting started process.
As for the wording, I quite like @ddemidovs proposed ' If you are just starting programming ev3dev, take a look at ev3python.com by @ndward. The site is intended to help users of the Lego EV3 Mindstorms to begin using the ev3dev Python programming language to control their robot.' But it seems to me that anyone coming to [ev3dev.org](url) probably already realises that the site is about writing programs to control the EV3 so maybe the text could be shortened to a single sentence: 'If you are a novice Python programmer then check out [ev3python.com](url) for help setting up and using EV3 Python'. (I've always called the language EV3 Python – is it better to call it EV3Dev Python?)

**Comment by WasabiFan** [comment]
At the top of this page, you should see tabs for "Conversation" "Commits" and "Changes". This thread is not an issue like the others you've seen, but actually it's the infamous "pull request". Essentially, it is a set of changes that @ddemidov proposed making to the site. If you click on the "Changes" tab you can see the changes that were included: he added a paragraph to the "programming languages" page (as you suggested) with a link to your site in it. At the top of the page, it also says "merged". This means that someone (in this case it was me) approved the changes and had those changes made in the main site code.

This means that the link to your site has already been added. You can see it here: http://www.ev3dev.org/docs/programming-languages/.

I can't speak to the language discussion in this thread, but it didn't have any effect on the choice to link to your site; it was just something related that people started discussion.

Also, don't worry about submitting your own pull request; we can make these changes, just keep in mind that it won't necessarily be immediate.

**Comment by ndward** [comment]
Thanks, @WasabiFan ! I look forward to the new pre-release!

File: docs/programming-languages.md [hunk 2]
```diff
@@ -17,6 +17,11 @@ Python is a high-level, general-purpose, interpreted scripting language. It is
 great for beginners, while also powerful for experienced coders.If you are new
```
Comment (bmegli) [inline docs/programming-languages.md]: missing space

now, this is a bit too late...

**Comment by ndward** [comment]
A link for anyone still interested in the Python vs Go debate. This 4 year old article is about professional-level coding and not very relevant to the discussion of what language beginning teens (most of whom will not become professional coders) should learn:
https://www.quora.com/What-advantages-does-the-Go-programming-language-have-over-Python-or-vice-versa

**Pull request merged**